### ISSUE TITLE
FLUID: Shell Command additional features

### ISSUE THREAD

**Issue by erco77** [fltk/fltk#674]
It would be best if fluid saved the `Shell -> Execute Command` in the .fl file so that the commands to build the .fl project are specific to that fluid project.

I'm not really sure where it stores it right now, but every time I change fluid files, I have to keep changing the commands around because it doesn't remember them in the .fl.

If hitting "save as default" saves the commands in a global place as the default for all new fluid projects, that's OK. But if the user changes the command, it should save that change in the .fl file, and each time that file is loaded, that should be the command that appears in the dialog.

From what I can tell the commands aren't saved in the .fl file.
There's controls in the dialog window I tried messing with, but I'm not sure exactly what they do:

- a "save as default" button
- a checkbox for "use settings in .fl design files"

These could really use some tooltip elaborations.

I'm left wondering what "save as default" does -- what's it saving and where is it saving it?

The "use settings in .fl files", I'm not sure which "settings" are being used, and why would those settings not be loaded from the .fl file, would they come from somewhere else?

I thought I understood the 'save as default' to save the commands I enter "somewhere", but they'd sometimes disappear within a fluid session, reverting to older/previous settings from older projects I haven't opened in a while. (Maybe ^Z undo operations affect this?)

And when I have the "use settings in .fl files" set, I was thinking it would save my commands to the .fl, but it doesn't seem to.

It would make more sense if _everything_ in the `Shell -> Execute Command` dialog were saved in the .fl file.

**Comment by Albrecht-S** [comment]
@MatthiasWM You asked for comments, so here are my general thoughts:
- I agree with Greg that it would be useful to store the shell commands (script) somewhere
- this should be done with respect to the particular project (`.fl`) file
- I also agree with Matthias that "importing" shell scripts with `.fl` files sounds dangerous
- ISTR that Matthias' first implementation was platform related, IIRC you could store commands related to the current platform (Windows, Unix/Linux, macOS)
- I don't think that the relation to a particular OS (platform) is very helpful. The user might want to store **different** commands on different **hosts** although the platform is the same.

In the following text I'm using "project" as a synonym for "`.fl` file".

My proposal is that shell scripts should be stored separate from project (`.fl`) files. If they were stored in (FLTK) user preferences we could rely on the **user** not storing hazardous commands in their own user preferences. The fluid user preferences could be stored together with the name or the path of the `.fl` file (or in a fluid prefs subdirectory, but that's an implementation detail).

Maybe a good approach would be to let the user store their shell commands not only in their user preferences **_per project_** but also in the **_fluid templates_** (in fluid prefs) so they can use "the same" (or similar) commands for different projects, i.e. load shell commands for a new or even an existing project from a stored template, change file names and paths according to their needs, and save them in their **_preferences for this project_**. This would also solve the "per platform" or "per host" configuration issues.

I wouldn't like the implementation of point 1 (just use whatever is in the .fl file). Also, storing the shell commands in the project file would make porting to or usage on different platforms or hosts more difficult (needs different per-host and/or per-platform commands), and **distribution of fluid files would become a nightmare**. Think of all the fluid files in the FLTK distro!

Regarding points 2 and 3: I wonder how to define **"_the first time_"**. If you compare the time stamp of the project file with the last confirmation, every change elsewhere in the file would prompt the user to confirm which would eventually make users confirm "automatically" - which renders the confirmation dialog useless and annoying. It would be possible to store the contents of the shell commands or a hash value together with or instead of the confirmation time [1], but that would be more complex.

Point 4 above is similar to what I propose.

Point 5: what if the user works on two different projects and changes often from one to the other project and back?

There may be more to it or better proposals, but this is what I could think of.

-------------------------
[1] All time-based confirmations are probably useless. What if the user unpacks a file from an archive (zip, tar, ...) and this sets the time stamp to a time before their latest confirmation? The only useful confirmation would likely need a valid hash of the contents to be safe, and then no confirmation time is needed at all, you'd just store the hash as confirmation in the user's preferences.

**Comment by MatthiasWM** [comment]
Yes, you are right about the hash value vs. the modification time.

Command line scripts are of course optional, and there is no need to save anything within the distribution project files. This is more of a hint for using a project file made by someone else. 99% of the time, the user will run "make" and not care.

So option number 6: project files can contain _suggested_ shell commands. They are not instated unless the user goes into Project Settings and actively copies the required commands to his host/platform/machine (could be a single click). The script commands are then stored elsewhere for that user on that platform. Updating the project file will not automatically update the chosen commands, and changing the commands will not update the in the project file either.

This would solve the unsolicited script issue (true opt-in), still give a useful script suggestion, allow host specific scripts (via preferences), and is minimal effort for the user.

**Comment by Albrecht-S** [comment]
> Command line scripts are of course optional, and there is no need to save anything within the distribution project files...

Sure, it's not necessary, but at some time a dev will certainly do it, maybe because it's a habit, or if he is working for a longer time on a project. Maybe a guy named Matthias working on fluid ;-)

Honestly, regarding point 6: what shell commands could one developer suggest others to use? Even if it's just `make` it would not be good for me, I'm always using `ninja`. And so on. The more I think about it, the more I come to the conclusion that storing any kind of user preferences (including shell commands) in fluid project files would be wrong.

If a project developer wanted to suggest a particular sequence of shell commands for their project, this should go into a README file and could then be copied by the user for their development. And the user could store this in their own preferences for further usage, as long as the project is developed.

I'm aware that storing setup (shell commands) and project info (fluid files) separately can have its own issues. You can leave stale shell commands in your preferences as soon as you stop working on a project and so on. But this can be solved somehow - for instance by storing the path to the fluid file in the preferences and letting the user purge stale entries. Another useful addition to fluid files might be a project ID (UUID) to link shell commands with projects. The UUID would then be stored in the .fl file. Again, this might be an issue if someone copied a fluid file for reuse in another project. Nothing is perfect. ;-)

**Comment by erco77** [comment]
Regarding the safety issue of shell commands in the .fl file..

- An .fl file could have a main() defined that invokes system("do a bad thing"), so you can already have a payload in a .fl file
- How is the Shell -> Execute any different than generating a Makefile that does bad things?
- Don't VS project files have custom commands? Should .fl be any different?

I just think the solving the problem domain of shell commands in an .fl file lies elsewhere.

**Comment by MatthiasWM** [comment]
See #328

**Comment by MatthiasWM** [comment]
Ok, so with commit 71b8e77935bc9b0a3afc97d322037f2549165d2f we are slowly coming to a conclusion on this feature.

There are still a hand full of features on my list:
 * [ ] save shell commands per directory in `./fluid.prefs`
 * [ ] save user defined shell commands per directory `./fluid.user.prefs`
 * [ ] add a variable `@FLTK_CONFIG@` that finds and runs `fltk-config`
 * [ ] add a bunch of sensible sample shell commands
 * [ ] add hostname, username, getenv support to conditional execution

**Comment by MatthiasWM** [comment]
This has been integrated into the new FLUID shell command handling

**Comment by MatthiasWM** [comment]
Not for 1.4.0, but maybe in 1.4.x

**Comment by MatthiasWM** [comment]
@erco77 I don`t remember where you mentioned that, but clearing the terminal history can be done with `echo "esc[3Jesc [ 2Jesc[H"`. On Unix, you can use `\x1b`. On MSWindows with cmd.exe, you need to insert the escape character with copy/paste from some other text editor (`Ctrl-[` in notepad++ works, IIRC).

**Comment by erco77** [comment]
@MatthiasWM Oh, you mean clearing the screen/hist before each shell execution?

Yes, it surprisingly did slip my mind I could put escape sequences in the Makefiles as a workaround, but that erases the nice little green message fluid prints showing the actual shell command being executed, which is useful for debugging esp. someone else's script, e.g.

![image](https://github.com/fltk/fltk/assets/6484779/a3846ca7-7260-405e-bb4c-6330d6b31a91)

When it comes to the use case of an IDE, fluid, building an app and going back to see the first error messages, it just seems like the most common use case is to clear before each execution so as not to be confused by old error messages from previous build attempts.

**Comment by MatthiasWM** [comment]
33d071875623dc90e8da00d0491f97bc2558c861 allow terminal clear and history clear 
![Screenshot 2023-12-16 at 13 53 43](https://github.com/fltk/fltk/assets/653993/945476db-41fe-43c8-b426-7ba180189aab)

**Comment by erco77** [comment]
Another great addition!

I would suggest tho: just have one "Clear terminal" button, and have it clear both the terminal and history.

The reason is the user doesn't see the screen actually clear with just "clear terminal" in this common use case:

- When the user scrolls back into a large (more than one page) compiler error report to see the first message, make fixes and hit the rebuild hotkey, the screen /does not clear/ because it's still scrolled back into the now old error message history, so the old compiler messages start scrolling by. The "display" area that was indeed cleared is so far below, it's not seen.

The other way to fix this is to, on screen clear, also set `tty->scrollbar->value(0)` so the scroll position is forced back to the bottom so the user is returned to the display area.

**Comment by MatthiasWM** [comment]
I opted for scrolling to the bottom. I find that more natural, and it's sometimes nice to have the history still available.

**Comment by MatthiasWM** [comment]
Moved into #1322
